From 1f9f50ea02164534c497f81356df7e6f1e89c2b4 Mon Sep 17 00:00:00 2001
From: Avi Niddam <aniddam@habana.ai>
Date: Fri, 06 Jun 2025 13:16:59 +0300
Subject: [PATCH] [SW-228856] fit: prepare original name back

Context:
We encountered several code duplications related to
fit_foreach_property.
So we created a new fit services infrastructure
to solve this issue in a clean way.
The main Lego pieces are the newly created files:
fit_services.c/h

Current commit:
In this current commit, we prepare the use back of the
original function name. This means: fit_foreach_property
instead of the temporary function _fit_foreach_property.

General remark:
This infrastructure is intended to support ANY of the
fit related functions that we find today in fit_utils.c
or fit.c
Therefore, from now, please, anyone doing a change related
to the fit files should carefully double-check:
Any unified code can be (and should be) located in
fit_services.c/h
For this, please, follow the example of how it has been
done with fit_foreach_property.

Change-Id: Ie521aa3aec829efbb2d0235ec73e5861efa28583
Signed-off-by: Avi Niddam <aniddam@habana.ai>
---

diff --git a/include/fit_services.h b/include/fit_services.h
index 266353c..4245c4d 100644
--- a/include/fit_services.h
+++ b/include/fit_services.h
@@ -18,4 +18,10 @@
 int _fit_foreach_property(const void *fit, const char *node_path,
 			_fit_property_match_cb match_cb, void *data);
 
+typedef int (*fit_property_match_cb)(const void *fit,
+		const struct fdt_property *prop, void *data);
+
+int fit_foreach_property(const void *fit, const char *node_path,
+			fit_property_match_cb match_cb, void *data);
+
 #endif
diff --git a/lib/common/fit_services.c b/lib/common/fit_services.c
index 23104a4..dc797eb 100644
--- a/lib/common/fit_services.c
+++ b/lib/common/fit_services.c
@@ -39,3 +39,31 @@
 
 	return count;
 }
+
+int fit_foreach_property(const void *fit, const char *node_path,
+			fit_property_match_cb match_cb, void *data)
+{
+	int ret, count = 0;
+	int noffset, property;
+	const struct fdt_property *prop;
+
+	noffset = fdt_path_offset(fit, node_path);
+	if (noffset < 0) {
+		EMB_LOG_RAW_ERR("Cannot find entry ");
+		EMB_LOG_RAW(node_path);
+		EMB_LOG_RAW("\n\r");
+		return noffset;
+	}
+
+	fdt_for_each_property_offset(property, fit, noffset) {
+		prop = fdt_get_property_by_offset(fit, property, NULL);
+		if (!prop)
+			continue;
+
+		ret = match_cb(fit, prop, data);
+		if (!ret)
+			count++;
+	}
+
+	return count;
+}
